collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java

```diff
@@ -210,25 +212,42 @@ public void uploadingComplete(HashMap<String, String> result) {
             // tried to close a dialog not open. don't care.
         }
 
-        StringBuilder selection = new StringBuilder();
+        StringBuilder selection = null;
         Set<String> keys = result.keySet();
         Iterator<String> it = keys.iterator();
 
-        String[] selectionArgs = new String[keys.size()];
-        int i = 0;
-        while (it.hasNext()) {
-            String id = it.next();
-            selection.append(InstanceColumns._ID + "=?");
-            selectionArgs[i++] = id;
-            if (i != keys.size()) {
-                selection.append(" or ");
+        String[] selectionArgs = null;
+        StringBuilder message = new StringBuilder();
+        int count = keys.size();
+
+        while (count != 0) {
+            int parLen = count;
+
+            if (count > SQLITE_MAXIMUM_QUERY_LIMIT) {
+                parLen = SQLITE_MAXIMUM_QUERY_LIMIT;
             }
-        }
 
-        StringBuilder message = new StringBuilder();
-        {
+            selection = new StringBuilder();
+            selectionArgs = new String[parLen];
+            selection.append(InstanceColumns._ID + " IN (");
+
+            int i = 0;
+            while (it.hasNext() && i < parLen) {
+                String id = it.next();
+                selectionArgs[i] = id;
+                if (i != parLen - 1) {
+                    selection.append("?,");
+                } else {
+                    selection.append("? )");
```

Comment:
And I think you can simplify this by always adding '?', and adding ',' only when parLen > 0 (the last '?' won't get a ','), and then outside of the of the while add the ')'. I think it's more readable that way. But that could be just  my preference.